cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -396,11 +397,11 @@ public void upgrade(boolean continueOnFailure) throws Exception {
           continue;
         }
 
-        Location appJarLocation = store.getApplicationArchiveLocation(appId);
+        Location appJarLocation = findAppJarLocation(appId);
         if (appJarLocation == null) {
-          LOG.error(String.format("Unable to get location of jar for app '%s' in namespace '%s'. " +
-              "You will need to re-deploy the app after upgrade.",
-            appId.getNamespaceId(), appId.getId()));
+          // nothing we can do... skip and log a message
+          LOG.warn("Unable to find application jar for app '%s' in namespace '%s'. " +
+            "Please re-deploy the app manually after upgrade.");
```

Comment:
Could throw an exception, but that would halt the upgrade and it would never be able to complete. I've opened CDAP-3551to surface these types of errors in a better way.